lib/spack/spack/test/conftest.py

```diff
@@ -1706,3 +1706,61 @@ def noncyclical_dir_structure(tmpdir):
         with open(j('file_3'), 'wb'):
             pass
     yield d
+
+
+@pytest.fixture(scope='function')
+def mock_url_config(monkeypatch):
+    """Mock web utility's curl and spider functions related to reading from
+       a URL that might represent a single configuration file or a directory
+       containing multiple configuration files.
+
+       Configuration information will be taken from the corresponding file(s)
+       in the test data configuration directory."""
+    config_data_dir = os.path.join(spack.paths.test_path, 'data', 'config')
+    config_files = os.listdir(config_data_dir)
+
+    def get_filename(url):
+        filename = os.path.basename(url)
+        if os.path.splitext(url)[1]:
+            return os.path.join(config_data_dir, filename) if filename \
+                in config_files else ''
+        return None
+
+    def get_file_urls(url):
+        # Any URL with an extension will be treated like a file; otherwise,
+        # it is considered a directory/folder and we'll grab all available
+        # files.
+        file_urls = []
+        if os.path.splitext(url)[1]:
+            file_urls.append(url)
+        else:
+            file_urls.extend([
+                '{0}{1}{2}'.format(url, os.sep, f) for f in config_files])
+        return file_urls
+
+    class MockCurl(object):
+        def __init__(self):
+            self.returncode = None
+
+        def __call__(self, *args, **kwargs):
+            url = [a for a in args if a.startswith('http')][0]
+            filename = get_filename(url)
+
+            with open(filename, 'r') as f:
+                lines = f.readlines()
+                write_file(os.path.basename(filename), ''.join(lines))
+
+            self.returncode = 0
+
+    def mock_curl(*args):
+        return MockCurl()
+
+    monkeypatch.setattr(spack.util.web, '_curl', mock_curl)
+
+    def mock_spider(*args, **kwargs):
+        urls = get_file_urls(args[0])
+        return [], set(urls)
```

Comment:
I think the combined mocking of `spider` and `curl` is intended to manufacture results for `fetch_remote_configs`: mocking the output of one function to be fed into another mocked function seems fragile: I think `fetch_remote_configs` should access this functionality via a single interface (e.g. one method that retrieves everything from a URL).

Reply:
I'm not sure I agree.  `fetch_remote_configs` has an inner function that fetches the raw contents of a (`.yaml`) file and is called to fetch text whether `fetch_remote_configs` is provided a URL to a 'file' or a 'directory'.  

`spack.util.web` grabs HTML for a URL and extracts associated links for it.  It is not practical IMO to extract the text from the HTML returned by spider.

So these are two separate functions with distinct purposes.

Would you prefer that I mock them separately instead of providing a single fixture?

Reply:
Or am I missing something?

Reply:
Is this all that different from, say, the `module_configuration` fixture where configuration settings are read from a test data file, set, and then retrieved through a function in a mock object?

Reply:
> Is this all that different from, say, the module_configuration fixture where configuration settings are read from a test data file, set, and then retrieved through a function in a mock object?

I assume we're talking about `tcl.test_autoload_direct` for example: what is the mocked object in that case? e.g. I don't think `modulefile_content` is mocking anything

(Overall I think this case is different, but I want to make sure I understand what you mean in the example)

> these are two separate functions with distinct purposes.

Apologies I don't mean that these functions need to be combined, but the code could add a single function that calls both of them, `fetch_remote_configs` could call that, and the test could likewise mock that function.

* One complication is that `fetch_remote_configs` also calls curl indirectly without `spider`, but I'll wait to think more about how to fix that until we reach a conclusion on whether this is actually a problem

Reply:
ping @scheibelp I'm still confused as indicated by my questions above.